core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterExchanger.java

```diff
@@ -0,0 +1,92 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.exchange;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+import io.trino.operator.exchange.PageReference.PageReleasedListener;
+import io.trino.spi.Page;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+public class ScaleWriterExchanger
+        implements LocalExchanger
+{
+    private static final Logger log = Logger.get(ScaleWriterExchanger.class);
+
+    private final List<Consumer<PageReference>> buffers;
+    private final LocalExchangeMemoryManager memoryManager;
+    private final PageReleasedListener onPageReleased;
+    private final long bufferSizeLimit;
+    private final Supplier<Long> physicalWrittenBytesSupplier;
+    private final long writerMinSize;
+
+    // start with single writer and increase the writer count as buffer memory utilization increases
+    @GuardedBy("this")
+    private int writerCount = 1;
+
+    @GuardedBy("this")
+    private int nextWriterIndex = -1;
+
+    public ScaleWriterExchanger(
+            List<Consumer<PageReference>> buffers,
+            LocalExchangeMemoryManager memoryManager,
+            long maxBufferedBytes,
+            Supplier<Long> physicalWrittenBytesSupplier,
+            DataSize writerMinSize)
+    {
+        this.buffers = requireNonNull(buffers, "buffers is null");
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.onPageReleased = PageReleasedListener.forLocalExchangeMemoryManager(memoryManager);
+        // We are aiming to keep the buffer size below 50% of its limit
+        this.bufferSizeLimit = maxBufferedBytes / 2;
+        this.physicalWrittenBytesSupplier = requireNonNull(physicalWrittenBytesSupplier, "physicalWrittenBytesSupplier is null");
+        this.writerMinSize = requireNonNull(writerMinSize, "writerMinSize is null").toBytes();
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Consumer<PageReference> buffer = buffers.get(getNextWriterIndex(memoryManager.getBufferedBytes(), physicalWrittenBytesSupplier.get()));
+        PageReference pageReference = new PageReference(page, 1, onPageReleased);
+        memoryManager.updateMemoryUsage(pageReference.getRetainedSizeInBytes());
+        buffer.accept(pageReference);
+    }
+
+    private synchronized int getNextWriterIndex(long bufferedBytes, long physicalWrittenBytes)
+    {
+        if (writerCount < buffers.size()
```

Comment:
When we have scaled up to max, we don't need to keep checking buffer size and physical written bytes.
So maybe something like below can be done
```
if (writerCount == buffers.size()) {
            synchronized (this) {
                nextWriterIndex = (nextWriterIndex + 1) % writerCount;
                return nextWriterIndex;
            }
        }
        long bufferedBytes = memoryManager.getBufferedBytes();
        long physicalWrittenBytes = physicalWrittenBytesSupplier.get();
        synchronized (this) {
            if (writerCount < buffers.size()
                    && bufferedBytes >= bufferSizeLimit
                    && physicalWrittenBytes >= writerCount * writerMinSize) {
                writerCount++;
                log.debug("Increased task writer count: %d", writerCount);
            }
            nextWriterIndex = (nextWriterIndex + 1) % writerCount;
            return nextWriterIndex;
        }
```

Reply:
I've removed the synchronized part since each source operator will have a separate instance of ScaleWriterExchanger.